### ISSUE TITLE
outage RCA April 24 2022  - API Actions, Sidekiq and Cron jobs

### ISSUE THREAD

**Issue by donrestarone** [restarone/violet_rails#478]
Servers became unresponsive after [this commit](https://github.com/restarone/violet_rails/commit/197b2fbeac5048325b63e172d7ca5394947bc570)

This was due to a variety of factors: 

1. Since jobs were running both in sidekiq and cron we ran into the thundering herd problem (see sidekiq's solutions to [it here](https://github.com/mperham/sidekiq/blob/main/Changes.md)) due to various workloads that were in the initialized state that were now being called every minute
2. As we fell back on jobs, systems started getting slower and slower until the OS crashed
3. When systems were booted back up (from forced restart) they functioned until sidekiq started processing jobs and some jobs were being rate limited (causing failed jobs) which then got picked up by cron and the server crashed due to just sheer workload. 
4. systems that had the API plugin `Subdomain Events (subdomain/event v1)` were hit especially hard since that plugin has a bug that spawns [multiple sidekiq jobs for a given event](https://github.com/restarone/violet_rails/issues/477) 

*TLDR; it was a workload bomb*

The fix was to either boot the server with the faulty code in place and turn off `sidekiq.service` or to deploy this [hotfix](https://github.com/restarone/violet_rails/commit/3377efc3fe7cd06493e06ecdc76f42fed43cbfbd)  

To ensure fastest time to recovery, the following steps were taken: 

1. Launch fresh instances from the latest stable AMI (faster time to recovery) -- to start serving web traffic (full performance was regained at 9.30AM EST)
2. after production systems were serving web traffic correctly debug validate assumptions and [deploy hotfix](https://github.com/restarone/violet_rails/commit/3377efc3fe7cd06493e06ecdc76f42fed43cbfbd) to staging (11AM EST)
3. verify for 1 hour on staging 
5. rotate public key secrets in the deployment pipeline
6. deploy to production environments (completed by 1PM EST)

**Comment by donrestarone** [comment]
Web performance is restored, but job performance keeps regressing and affecting EC2 instance health. Our crontab is as follows: 

``` txt
Last login: Sun Apr 24 18:24:39 2022 from 76.64.37.224
ubuntu@ip-172-31-86-57:~$ crontab -l


# Begin Whenever generated tasks for: violet at: 2022-04-24 17:18:35 +0000
PATH=/var/www/violet/shared/bundle/ruby/2.6.0/bin:/home/ubuntu/.rbenv/versions/2.6.6/bin:/home/ubuntu/.rbenv/libexec:/home/ubuntu/.rbenv/plugins/ruby-build/bin:/home/ubuntu/.rbenv/plugins/rbenv-vars/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin

0 0,12 * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake maintenance:clear_old_ahoy_visits --silent >> log/whenever.log 2>&1'

0 0,12 * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=twelve_hours --silent >> log/whenever.log 2>&1'

* * * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=every_minute --silent >> log/whenever.log 2>&1'

* * * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake api_action:run_initialized_actions --silent >> log/whenever.log 2>&1'

0,5,10,15,20,25,30,35,40,45,50,55 * * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=five_minutes --silent >> log/whenever.log 2>&1'

0,5,10,15,20,25,30,35,40,45,50,55 * * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake api_action:rerun_failed_actions --silent >> log/whenever.log 2>&1'

0,10,20,30,40,50 * * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=ten_minutes --silent >> log/whenever.log 2>&1'

0,30 * * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=thirty_minutes --silent >> log/whenever.log 2>&1'

0 * * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=every_hour --silent >> log/whenever.log 2>&1'

0 0,3,6,9,12,15,18,21 * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=three_hours --silent >> log/whenever.log 2>&1'

0 0,6,12,18 * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=six_hours --silent >> log/whenever.log 2>&1'

0 0 * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=one_day --silent >> log/whenever.log 2>&1'

0 0 * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake report:send_analytics_report --silent >> log/whenever.log 2>&1'

0 0 * * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake maintenance:clear_discarded_api_actions --silent >> log/whenever.log 2>&1'

0 0 1,8,15,22 * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=one_week --silent >> log/whenever.log 2>&1'

0 0 1,15 * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=two_weeks --silent >> log/whenever.log 2>&1'

0 0 1 * * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=one_month --silent >> log/whenever.log 2>&1'

0 0 1 1,4,7,10 * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=three_months --silent >> log/whenever.log 2>&1'

0 0 1 1,7 * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=six_months --silent >> log/whenever.log 2>&1'

0 0 1 1 * /bin/bash -l -c 'export PATH="$HOME/.rbenv/bin:$PATH"; eval "$(rbenv init -)"; cd /var/www/violet/releases/20220424171545 && RAILS_ENV=staging bundle exec rake external_api_client:drive_cron_jobs CRON_INTERVAL=one_year --silent >> log/whenever.log 2>&1'

# End Whenever generated tasks for: violet at: 2022-04-24 17:18:35 +0000
ubuntu@ip-172-31-86-57:~$
```

Looking at it, the only mission critical jobs are the ones related to running failed api actions + reporting tasks. To debug the issue with jobs falling back and killing EC2 instances, I have temporarily removed the crontabs so while we wont be processing those jobs we will at least have stable web performance to buy us some time.

**Comment by donrestarone** [comment]
Update after removing the cron tabs, 

Some instances still became unresponsive. From here we have 2 strategies to narrow down the issue:

- go back, commit by commit on master until we find the point in time where the issue was introduced 
- assume the issue is with libraries-- and setup a fresh AMI and test

**Comment by donrestarone** [comment]
The following commits on master have been deemed stable for production use:

1. Branch master (at 5642e8c521a3353bf391efeadeb5bcbddfd018bb) - validated on markedrestaurant.com
2. Branch master (at 3377efc3fe7cd06493e06ecdc76f42fed43cbfbd) - validated on nikean.org
3. Branch master (at 11548a1a7cdb136467e81604c5a65a0947d99ecb) - validated on sipshucksip.com
4. Branch master (at 229cd1feb5fb6f259210acf358ca58c12a5eed1f) - validated on sanjaysinghal.com
5. Branch master (at 3377efc3fe7cd06493e06ecdc76f42fed43cbfbd) - validated on atoi.to

CC @alis-khadka @Pralish

**Comment by donrestarone** [comment]
we are now testing app specific cron jobs. Looks like an externalAPIClient is overloading the system

![Screenshot from 2022-04-28 20-01-26](https://user-images.githubusercontent.com/35935196/165866325-f0e7dcf8-b679-4603-82e5-d6ae71fced57.png)

**Comment by donrestarone** [comment]
prevented the same externalApiClient from running at the same time

https://github.com/restarone/violet_rails/pull/496

but instances are still becoming unhealthy. Turning off cron on staging but still running the latest commit [here](https://github.com/restarone/violet_rails/pull/496/commits/5fce798d0853cd5065033698a0e27d655a00e060) -- lets check back after 24 hours to see if the system is stable with the latest changes and cron turned off
